nutils/cli.py

```diff
@@ -14,82 +14,69 @@ def _githash( path, depth=0  ):
     githash, = ref.read().split()
   return githash
 
-def getkwargdefaults( func ):
-  'helper for run'
+def _bool( s ):
+  if s in ('true','True'):
+    return True
+  if s in ('false','False'):
+    return False
+  raise argparse.ArgumentTypeError( 'invalid boolean value: {!r}'.format(s) )
 
-  kwargs = util.OrderedDict()
+def _create_function_parser( func ):
+  parser = argparse.ArgumentParser( usage=argparse.SUPPRESS, formatter_class=argparse.ArgumentDefaultsHelpFormatter, add_help=False )
+  parser._optionals.title = 'optional arguments for ' + func.__name__ # NOTE undocumented
   signature = inspect.signature( func )
   for parameter in signature.parameters.values():
-    if parameter.kind in (parameter.VAR_POSITIONAL, parameter.VAR_KEYWORD):
-      continue
-    if parameter.default is parameter.empty:
-      raise ValueError( 'Function cannot be called without arguments.' )
-    kwargs[parameter.name] = parameter.default
-  return kwargs
+    parser.add_argument(
+      '--'+parameter.name,
+      type=str,
+      default=parameter.default,
+      metavar=parameter.name[0].upper(),
+      help=parameter.annotation if parameter.annotation is not inspect._empty else None,
+    )
+  return parser
+
+def _parse_args( functions ):
+  parser = argparse.ArgumentParser( usage='%(prog)s [{}] [--arg=X [...]]'.format( '|'.join(func.__name__ for func in functions) ) )
```

Comment:
Minor thing: to be in line with many unix utilities (cp, mv, ls, ...) I'd suggest `'%(prog)s [{}] [OPTION]...'` as usage string.